components/suggest/src/util.rs

```diff
@@ -0,0 +1,1578 @@
+/// paths that don't start with `"a"`:
+///
+/// ```
+/// # use suggest::util::filter_map_chunks;
+/// let paths = filter_map_chunks("a b c", 3, |chunk, chunk_index, _| {
+///     if chunk_index > 0 || chunk == "a" {
+///         Ok(Some(vec![chunk.to_uppercase()]))
+///     } else {
+///         Ok(None)
+///     }
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["A", "B", "C"],
+///     vec!["A", "B C"],
+/// ]);
+/// ```
+///
+/// Same as the first example but using `path` in the filter-map function to
+/// prune paths that include "A B":
+///
+/// ```
+/// # use suggest::util::filter_map_chunks;
+/// let paths = filter_map_chunks("a b c", 3, |chunk, _, path| {
+///     if path.iter().any(|value| value == "A B") {
+///         Ok(None)
+///     } else {
+///         Ok(Some(vec![chunk.to_uppercase()]))
+///     }
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["A", "B", "C"],
+///     vec!["A", "B C"],
+///     vec!["A B C"],
+/// ]);
+/// ```
+///
+/// Mapping each chunk to multiple values:
+///
+/// ```
+/// # use suggest::util::filter_map_chunks;
+/// let paths = filter_map_chunks("a b c", 3, |chunk, _, _| {
+///     Ok(Some(vec![format!("{chunk}0"), format!("{chunk}1")]))
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["a0", "b0", "c0"],
+///     vec!["a0", "b0", "c1"],
+///     vec!["a0", "b1", "c0"],
+///     vec!["a0", "b1", "c1"],
+///     vec!["a0", "b c0"],
+///     vec!["a0", "b c1"],
+///     vec!["a1", "b0", "c0"],
+///     vec!["a1", "b0", "c1"],
+///     vec!["a1", "b1", "c0"],
+///     vec!["a1", "b1", "c1"],
+///     vec!["a1", "b c0"],
+///     vec!["a1", "b c1"],
+///     vec!["a b0", "c0"],
+///     vec!["a b0", "c1"],
+///     vec!["a b1", "c0"],
+///     vec!["a b1", "c1"],
+///     vec!["a b c0"],
+///     vec!["a b c1"]
+/// ]);
+/// ```
+pub fn filter_map_chunks<T: Clone>(
+    query: &str,
+    max_chunk_size: usize,
+    f: impl Fn(String, usize, &[T]) -> Result<Option<Vec<T>>>,
+) -> Result<Vec<Vec<T>>> {
+    let words: Vec<_> = query.split_whitespace().filter(|s| !s.is_empty()).collect();
+    filter_map_chunks_recurse(&words, max_chunk_size, &f, vec![], 0)
+}
+
+fn filter_map_chunks_recurse<T: Clone>(
+    words: &[&str],
+    max_chunk_size: usize,
+    f: &impl Fn(String, usize, &[T]) -> Result<Option<Vec<T>>>,
+    path: Vec<T>,
+    chunk_index: usize,
+) -> Result<Vec<Vec<T>>> {
+    // Filtered-in (non-pruned) paths that will be returned from this step of
+    // the traversal.
+    let mut this_step_paths: Vec<Vec<T>> = vec![];
+
+    for chunk_size in 1..=max_chunk_size {
+        if words.len() < chunk_size {
+            // `chunk_size` and the later chunk sizes in this for-loop are too
+            // big to visit the remaining words. We already visited them earlier
+            // in the loop when the chunk size was small enough.
+            break;
+        }
+
+        let chunk = words[..chunk_size].join(" ");
```

Comment:
There's probably a way to do this with using `&str` objects to avoid allocating a string for each join.

Reply:
Hmm that might be doable. We could pass around the original query string instead of or probably in addition to the `words` slice, and then get the candidate chunk as a slice of the original string. I'll try that.

Reply:
That worked. I was also able to avoid the `path.clone()` by passing around `&mut path` and popping mapped values from it after recursing.